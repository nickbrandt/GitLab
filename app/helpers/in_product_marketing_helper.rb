# frozen_string_literal: true

module InProductMarketingHelper
  def subject_line(track, series)
    {
      create: [
        s_('InProductMarketing|Create a project in GitLab in 5 minutes'),
        s_('InProductMarketing|Import your project and code from GitHub, Bitbucket and others'),
        s_('InProductMarketing|Understand repository mirroring')
      ],
      verify: [
        s_('InProductMarketing|Feel the need for speed?'),
        s_('InProductMarketing|3 ways to dive into GitLab CI/CD'),
        s_('InProductMarketing|Explore the power of GitLab CI/CD')
      ],
      trial: [
        s_('InProductMarketing|Go farther with GitLab'),
        s_('InProductMarketing|Automated security scans directly within GitLab'),
        s_('InProductMarketing|Take your source code management to the next level')
      ],
      team: [
        s_('InProductMarketing|Working in GitLab = more efficient'),
        s_("InProductMarketing|Multiple owners, confusing workstreams? We've got you covered"),
        s_('InProductMarketing|Your teams can be more efficient')
      ]
    }[track][series]
  end

  def in_product_marketing_logo(track, series)
    inline_image_link('mailers/in_product_marketing', "#{track}-#{series}.png", { width: '150', style: 'width: 150px;' })
  end

  def about_link(folder, image, width)
    link_to inline_image_link(folder, image, { width: width, style: "width: #{width}px;", alt: s_('InProductMarketing|go to about.gitlab.com') }), 'https://about.gitlab.com/'
  end

  def in_product_marketing_tagline(track, series)
    {
      create: [
        s_('InProductMarketing|Get started today'),
        s_('InProductMarketing|Get our import guides'),
        s_('InProductMarketing|Need an alternative to importing?')
      ],
      verify: [
        s_('InProductMarketing|Use GitLab CI/CD'),
        s_('InProductMarketing|Test, create, deploy'),
        s_('InProductMarketing|Are your runners ready?')
      ],
      trial: [
        s_('InProductMarketing|Start a free trial of GitLab Ultimate – no CC required'),
        s_('InProductMarketing|Improve app security with a 30-day trial'),
        s_('InProductMarketing|Start with a GitLab Gold free trial')
      ],
      team: [
        s_('InProductMarketing|Invite your colleagues to join in less than one minute'),
        s_('InProductMarketing|Get your team set up on GitLab'),
        nil
      ]
    }[track][series]
  end

  def in_product_marketing_title(track, series)
    {
      create: [
        s_('InProductMarketing|Take your first steps with GitLab'),
        s_('InProductMarketing|Start by importing your projects'),
        s_('InProductMarketing|How (and why) mirroring makes sense')
      ],
      verify: [
        s_('InProductMarketing|Rapid development, simplified'),
        s_('InProductMarketing|Get started with GitLab CI/CD'),
        s_('InProductMarketing|Launch GitLab CI/CD in 20 minutes or less')
      ],
      trial: [
        s_('InProductMarketing|Give us one minute...'),
        s_("InProductMarketing|Security that's integrated into your development lifecycle"),
        s_('InProductMarketing|Improve code quality and streamline reviews')
      ],
      team: [
        s_('InProductMarketing|Team work makes the dream work'),
        s_('InProductMarketing|*GitLab*, noun: a synonym for efficient teams'),
        s_('InProductMarketing|Find out how your teams are really doing')
      ]
    }[track][series]
  end

  def in_product_marketing_subtitle(track, series)
    {
      create: [
        s_('InProductMarketing|Dig in and create a project and a repo'),
        s_("InProductMarketing|Here's what you need to know"),
        s_('InProductMarketing|Try it out')
      ],
      verify: [
        s_('InProductMarketing|How to build and test faster'),
        s_('InProductMarketing|Explore the options'),
        s_('InProductMarketing|Follow our steps')
      ],
      trial: [
        s_('InProductMarketing|...and you can get a free trial of GitLab Gold'),
        s_('InProductMarketing|Try GitLab Gold for free'),
        s_('InProductMarketing|Better code in less time')
      ],
      team: [
        s_('InProductMarketing|Actually, GitLab makes the team work (better)'),
        s_('InProductMarketing|Our tool brings all the things together'),
        s_("InProductMarketing|It's all in the stats")
      ]
    }[track][series]
  end

  def in_product_marketing_body_line1(track, series, format: nil)
    {
      create: [
        s_("InProductMarketing|To understand and get the most out of GitLab, start at the beginning and %{project_link}. In GitLab, repositories are part of a project, so after you've created your project you can go ahead and %{repo_link}.") % { project_link: project_link(format), repo_link: repo_link(format) },
        s_("InProductMarketing|Making the switch? It's easier than you think to import your projects into GitLab. Move %{github_link}, or import something %{bitbucket_link}.") % { github_link: github_link(format), bitbucket_link: bitbucket_link(format) },
        s_("InProductMarketing|Sometimes you're not ready to make a full transition to a new tool. If you're not ready to fully commit, %{mirroring_link} gives you a safe way to try out GitLab in parallel with your current tool.") % { mirroring_link: mirroring_link(format) }
      ],
      verify: [
        s_("InProductMarketing|Tired of wrestling with disparate tool chains, information silos and inefficient processes? GitLab's CI/CD is built on a DevOps platform with source code management, planning, monitoring and more ready to go. Find out %{ci_link}.") % { ci_link: ci_link(format) },
        s_("InProductMarketing|GitLab's CI/CD makes software development easier. Don't believe us? Here are three ways you can take it for a fast (and satisfying) test drive:"),
        s_("InProductMarketing|Get going with CI/CD quickly using our %{quick_start_link}. Start with an available runner and then create a CI .yml file – it's really that easy.") % { quick_start_link: quick_start_link(format) }
      ],
      trial: [
        [
          s_("InProductMarketing|GitLab's premium tiers are designed to make you, your team and your application more efficient and more secure with features including but not limited to:"),
          list([
            s_('InProductMarketing|%{strong_start}Company wide portfolio management%{strong_end} — including multi-level epics, scoped labels').html_safe % strong_options(format),
            s_('InProductMarketing|%{strong_start}Multiple approval roles%{strong_end} — including code owners and required merge approvals').html_safe % strong_options(format),
            s_('InProductMarketing|%{strong_start}Advanced application security%{strong_end} — including SAST, DAST scanning, FUZZ testing, dependency scanning, license compliance, secrete detection').html_safe % strong_options(format),
            s_('InProductMarketing|%{strong_start}Executive level insights%{strong_end} — including reporting on productivity, tasks by type, days to completion, value stream').html_safe % strong_options(format)
          ], format)
        ].join("\n"),
        s_('InProductMarketing|GitLab provides static application security testing (SAST), dynamic application security testing (DAST), container scanning, and dependency scanning to help you deliver secure applications along with license compliance.'),
        s_('InProductMarketing|By enabling code owners and required merge approvals the right person will review the right MR. This is a win-win: cleaner code and a more efficient review process.')
      ],
      team: [
        [
          s_('InProductMarketing|Did you know teams that use GitLab are far more efficient?'),
          list([
            s_('InProductMarketing|Goldman Sachs went from 1 build every two weeks to thousands of builds a day'),
            s_('InProductMarketing|Ticketmaster decreased their CI build time by 15X')
          ], format)
        ].join("\n"),
        s_("InProductMarketing|We know a thing or two about efficiency and we don't want to keep that to ourselves. Sign up for a free trial of GitLab Gold and your teams will be on it from day one."),
        [
          s_('InProductMarketing|Stop wondering and use GitLab to answer questions like:'),
          list([
            s_('InProductMarketing|How long does it take us to close issues/MRs by types like feature requests, bugs, tech debt, security?'),
            s_('InProductMarketing|How many days does it take our team to complete various tasks?'),
            s_('InProductMarketing|What does our value stream timeline look like from product to development to review and production?')
          ], format)
        ].join("\n")
      ]
    }[track][series]
  end

  def in_product_marketing_body_line2(track, series, format: nil)
    {
      create: [
        s_("InProductMarketing|That's all it takes to get going with GitLab, but if you're new to working with Git, check out our %{basics_link} for helpful tips and tricks for getting started.") % { basics_link: basics_link(format) },
        s_("InProductMarketing|Have a different instance you'd like to import? Here's our %{import_link}.") % { import_link: import_link(format) },
        s_("InProductMarketing|It's also possible to simply %{external_repo_link} in order to take advantage of GitLab's CI/CD.") % { external_repo_link: external_repo_link(format) }
      ],
      verify: [
        nil,
        list([
          s_('InProductMarketing|Start by %{performance_link}').html_safe % { performance_link: performance_link(format) },
          s_('InProductMarketing|Move on to easily creating a Pages website %{ci_template_link}').html_safe % { ci_template_link: ci_template_link(format) },
          s_('InProductMarketing|And finally %{deploy_link} a Python application.').html_safe % { deploy_link: deploy_link(format) }
        ], format),
        nil
      ],
      trial: [
        s_('InProductMarketing|Start a GitLab Gold trial today in less than one minute, no credit card required.'),
        s_('InProductMarketing|Get started today with a 30-day GitLab Gold trial, no credit card required.'),
        s_('InProductMarketing|Code owners and required merge approvals are part of the paid tiers of GitLab. You can start a free 30-day trial of GitLab Gold and enable these features in less than 5 minutes with no credit card required.')
      ],
      team: [
        s_('InProductMarketing|Invite your colleagues and start shipping code faster.'),
        s_("InProductMarketing|Streamline code review, know at a glance who's unavailable, communicate in comments or in email and integrate with Slack so everyone's on the same page."),
        s_('InProductMarketing|When your team is on GitLab these answers are a click away.')
      ]
    }[track][series]
  end

  def cta_link(track, series, group, format: nil)
    case format
    when :html
      link_to in_product_marketing_cta_text(track, series), group_email_campaigns_url(group, track: track, series: series), target: '_blank', rel: 'noopener noreferrer'
    else
      [in_product_marketing_cta_text(track, series), group_email_campaigns_url(group, track: track, series: series)].join(' >> ')
    end
  end

  def in_product_marketing_progress(track, series)
    s_('InProductMarketing|This is email %{series} of 3 in the %{track} series.') % { series: series + 1, track: track.to_s.humanize }
  end

  def footer_links(format: nil)
    links = [
      [s_('InProductMarketing|Blog'), 'https://about.gitlab.com/blog'],
      [s_('InProductMarketing|Twitter'), 'https://twitter.com/gitlab'],
      [s_('InProductMarketing|Facebook'), 'https://www.facebook.com/gitlab'],
      [s_('InProductMarketing|YouTube'), 'https://www.youtube.com/channel/UCnMGQ8QHMAnVIsI3xJrihhg']
    ]
    case format
    when :html
      links.map do |text, link|
        link_to(text, link)
      end
    else
      '| ' + links.map do |text, link|
        [text, link].join(' ')
      end.join("\n| ")
    end
  end

  def address(format: nil)
    s_('InProductMarketing|%{strong_start}GitLab Inc.%{strong_end} 268 Bush Street, #350, San Francisco, CA 94104, USA').html_safe % strong_options(format)
  end

  def unsubscribe(format: nil)
    parts = [
      s_('InProductMarketing|If you no longer wish to receive marketing emails from us,'),
      s_('InProductMarketing|you may %{unsubscribe_link} at any time.') % { unsubscribe_link: unsubscribe_link(format) }
    ]
    case format
    when :html
      parts.join(' ')
    else
      parts.join("\n" + ' ' * 16)
    end
  end

  private

  def in_product_marketing_cta_text(track, series)
    {
      create: [
        s_('InProductMarketing|Create your first project!'),
        s_('InProductMarketing|Master the art of importing!'),
        s_('InProductMarketing|Understand your project options')
      ],
      verify: [
        s_('InProductMarketing|Get to know GitLab CI/CD'),
        s_('InProductMarketing|Try it yourself'),
        s_('InProductMarketing|Explore GitLab CI/CD')
      ],
      trial: [
        s_('InProductMarketing|Start a trial'),
        s_('InProductMarketing|Beef up your security'),
        s_('InProductMarketing|Go for the gold!')
      ],
      team: [
        s_('InProductMarketing|Invite your colleagues today'),
        s_('InProductMarketing|Invite your team in less than 60 seconds'),
        s_('InProductMarketing|Invite your team now')
      ]
    }[track][series]
  end

  def project_link(format)
    link(s_('InProductMarketing|create a project'), help_page_url('gitlab-basics/create-project'), format)
  end

  def repo_link(format)
    link(s_('InProductMarketing|set up a repo'), help_page_url('user/project/repository/index', anchor: 'create-a-repository'), format)
  end

  def github_link(format)
    link(s_('InProductMarketing|GitHub Enterprise projects to GitLab'), help_page_url('integration/github'), format)
  end

  def bitbucket_link(format)
    link(s_('InProductMarketing|from Bitbucket'), help_page_url('user/project/import/bitbucket_server'), format)
  end

  def mirroring_link(format)
    link(s_('InProductMarketing|repository mirroring'), help_page_url('user/project/repository/repository_mirroring'), format)
  end

  def ci_link(format)
    link(s_('InProductMarketing|how easy it is to get started'), help_page_url('ci/README'), format)
  end

  def performance_link(format)
    link(s_('InProductMarketing|testing browser performance'), help_page_url('user/project/merge_requests/browser_performance_testing'), format)
  end

  def ci_template_link(format)
    link(s_('InProductMarketing|using a CI/CD template'), help_page_url('user/project/pages/getting_started/pages_ci_cd_template'), format)
  end

  def deploy_link(format)
    link(s_('InProductMarketing|test and deploy'), help_page_url('ci/examples/test-and-deploy-python-application-to-heroku'), format)
  end

  def quick_start_link(format)
    link(s_('InProductMarketing|quick start guide'), help_page_url('ci/quick_start/README'), format)
  end

  def basics_link(format)
    link(s_('InProductMarketing|Git basics'), help_page_url('gitlab-basics/README'), format)
  end

  def import_link(format)
    link(s_('InProductMarketing|comprehensive guide'), help_page_url('user/project/import/index'), format)
  end

  def external_repo_link(format)
    link(s_('InProductMarketing|connect an external repository'), new_project_url(anchor: 'cicd_for_external_repo'), format)
  end

  def unsubscribe_link(format)
    link(s_('InProductMarketing|unsubscribe'), '%tag_unsubscribe_url%', format)
  end

  def link(text, link, format)
    case format
    when :html
      link_to text, link
    else
      "#{text} (#{link})"
    end
  end

  def list(array, format)
    case format
    when :html
      tag.ul { array.map { |item| concat tag.li item} }
    else
      '- ' + array.join("\n- ")
    end
  end

  def strong_options(format)
    case format
    when :html
      { strong_start: '<b>'.html_safe, strong_end: '</b>'.html_safe }
    else
      { strong_start: '', strong_end: '' }
    end
  end

  def inline_image_link(folder, image, options)
    attachments.inline[image] = File.read(Rails.root.join("app/assets/images", folder, image))
    image_tag attachments[image].url, **options
  end
end
