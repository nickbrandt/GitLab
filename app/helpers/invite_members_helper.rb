# frozen_string_literal: true

module InviteMembersHelper
  include Gitlab::Utils::StrongMemoize

  def can_invite_members_for_project?(project)
    Feature.enabled?(:invite_members_group_modal, project.group) && can_manage_project_members?(project)
  end

  def can_invite_group_for_project?(project)
    Feature.enabled?(:invite_members_group_modal, project.group) && can_manage_project_members?(project) && project.allowed_to_share_with_group?
  end

  def directly_invite_members?
    strong_memoize(:directly_invite_members) do
      can_import_members?
    end
  end

  def invite_group_members?(group)
    experiment_enabled?(:invite_members_empty_group_version_a) && Ability.allowed?(current_user, :admin_group_member, group)
  end

  def invite_accepted_notice(member)
    case member.source
    when Project
      _("You have been granted %{member_human_access} access to project %{name}.") %
        { member_human_access: member.human_access, name: member.source.name }
    when Group
      _("You have been granted %{member_human_access} access to group %{name}.") %
        { member_human_access: member.human_access, name: member.source.name }
    end
  end
end
