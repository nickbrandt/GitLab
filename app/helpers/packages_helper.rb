# frozen_string_literal: true

module PackagesHelper
  def package_sort_path(options = {})
    "#{request.path}?#{options.to_param}"
  end

  def nuget_package_registry_url(project_id)
    expose_url(api_v4_projects_packages_nuget_index_path(id: project_id, format: '.json'))
  end

  def package_registry_instance_url(registry_type)
    expose_url("api/#{::API::API.version}/packages/#{registry_type}")
  end

  def package_registry_project_url(project_id, registry_type = :maven)
    project_api_path = expose_path(api_v4_projects_path(id: project_id))
    package_registry_project_path = "#{project_api_path}/packages/#{registry_type}"
    expose_url(package_registry_project_path)
  end

  def package_from_presenter(package, user)
    presenter = ::Packages::Detail::PackagePresenter.new(package)
    presenter.detail_view(user).to_json
  end

  def pypi_registry_url(project_id)
    full_url = expose_url(api_v4_projects_packages_pypi_simple_package_name_path({ id: project_id, package_name: '' }, true))
    full_url.sub!('://', '://__token__:<your_personal_token>@')
  end

  def composer_registry_url(group_id)
    expose_url(api_v4_group___packages_composer_packages_path(id: group_id, format: '.json'))
  end

  def composer_config_repository_name(group_id)
    "#{Gitlab.config.gitlab.host}/#{group_id}"
  end

  def packages_list_data(type, resource)
    {
      resource_id: resource.id,
      page_type: type,
      empty_list_help_url: help_page_path('user/packages/package_registry/index'),
      empty_list_illustration: image_path('illustrations/no-packages.svg'),
      package_help_url: help_page_path('user/packages/index')
    }
  end

  def track_package_event(event_name, scope, **args)
    ::Packages::CreateEventService.new(nil, current_user, event_name: event_name, scope: scope).execute
    category = args.delete(:category) || self.class.name
    ::Gitlab::Tracking.event(category, event_name.to_s, **args)
  end
end
