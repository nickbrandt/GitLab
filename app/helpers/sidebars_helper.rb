# frozen_string_literal: true

module SidebarsHelper
  def sidebar_tracking_attributes_by_object(object)
    case object
    when Project
      sidebar_project_tracking_attrs
    when Group
      sidebar_group_tracking_attrs
    when User
      sidebar_user_profile_tracking_attrs
    else
      {}
    end
  end

  def project_sidebar_context(project, user, current_ref)
    context_data = project_sidebar_context_data(project, user, current_ref)

    Sidebars::Projects::Context.new(**context_data)
  end

  private

  def sidebar_project_tracking_attrs
    tracking_attrs('projects_side_navigation', 'render', 'projects_side_navigation')
  end

  def sidebar_group_tracking_attrs
    tracking_attrs('groups_side_navigation', 'render', 'groups_side_navigation')
  end

  def sidebar_user_profile_tracking_attrs
    tracking_attrs('user_side_navigation', 'render', 'user_side_navigation')
  end

  def project_sidebar_context_data(project, user, current_ref)
    {
      current_user: user,
      container: project,
      learn_gitlab_experiment_enabled: learn_gitlab_experiment_enabled?(project),
      current_ref: current_ref,
      jira_issues_integration: project_jira_issues_integration?,
      can_view_pipeline_editor: can_view_pipeline_editor?(project)
    }
  end
end
