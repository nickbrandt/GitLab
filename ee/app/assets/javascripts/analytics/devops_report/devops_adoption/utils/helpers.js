import { isToday } from '~/lib/utils/datetime_utility';

/**
 * A helper function which accepts the enabledNamespaces,
 *
 * @param {Object} params the enabledNamespaces data, timestamp and check for open modals
 *
 * @return {Boolean} a boolean to determine if table data should be polled
 */
export const shouldPollTableData = ({ enabledNamespaces, timestamp, openModal }) => {
  if (openModal) {
    return false;
  } else if (!enabledNamespaces.length) {
    return true;
  }

  const anyPendingEnabledNamespaces = enabledNamespaces.some(
    (node) => node.latestSnapshot === null,
  );
  const dataNotRefreshedToday = !isToday(new Date(timestamp));

  return anyPendingEnabledNamespaces || dataNotRefreshedToday;
};
