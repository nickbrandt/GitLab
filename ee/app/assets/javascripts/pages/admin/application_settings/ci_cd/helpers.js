/**
 * Filters [gitlabCiYmls] based on a given [searchTerm].
 * Yml catagories with no items after filtering are not included in the returned object.
 * @param {Object} gitlabCiYmls - { <categoryName>: [{ name, id }] }
 * @param {String} searchTerm
 * @returns {Object}
 */
export function filterGitlabCiYmls(gitlabCiYmls, searchTerm) {
  return Object.keys(gitlabCiYmls).reduce((filteredYmls, category) => {
    const categoryYmls = gitlabCiYmls[category].filter((yml) =>
      yml.name.toLowerCase().startsWith(searchTerm),
    );

    if (categoryYmls.length > 0) {
      Object.assign(filteredYmls, {
        [category]: categoryYmls,
      });
    }

    return filteredYmls;
  }, {});
}
