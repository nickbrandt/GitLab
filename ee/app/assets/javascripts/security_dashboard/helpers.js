import isPlainObject from 'lodash/isPlainObject';
import { REPORT_TYPES, SEVERITY_LEVELS } from 'ee/security_dashboard/store/constants';
import { BASE_FILTERS } from 'ee/security_dashboard/store/modules/filters/constants';
import convertReportType from 'ee/vue_shared/security_reports/store/utils/convert_report_type';
import { VULNERABILITY_STATES } from 'ee/vulnerabilities/constants';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { s__, __ } from '~/locale';
import { DEFAULT_SCANNER } from './constants';

const parseOptions = (obj) =>
  Object.entries(obj).map(([id, name]) => ({ id: id.toUpperCase(), name }));

export const mapProjects = (projects) =>
  projects.map((p) => ({ id: p.id.split('/').pop(), name: p.name }));

const stateOptions = parseOptions(VULNERABILITY_STATES);
const defaultStateOptions = stateOptions.filter((x) => ['DETECTED', 'CONFIRMED'].includes(x.id));

export const stateFilter = {
  name: s__('SecurityReports|Status'),
  id: 'state',
  options: stateOptions,
  allOption: BASE_FILTERS.state,
  defaultOptions: defaultStateOptions,
};

export const severityFilter = {
  name: s__('SecurityReports|Severity'),
  id: 'severity',
  options: parseOptions(SEVERITY_LEVELS),
  allOption: BASE_FILTERS.severity,
  defaultOptions: [],
};

export const createScannerOption = (vendor, reportType) => {
  const type = reportType.toUpperCase();

  return {
    id: `${vendor}.${type}`,
    reportType: reportType.toUpperCase(),
    name: convertReportType(reportType),
    scannerIds: [],
  };
};

// This is used on the pipeline security tab, group-level report, and instance-level report. It's
// used by the scanner filter that shows a flat list of scan types (DAST, SAST, etc) with no vendor
// grouping.
export const simpleScannerFilter = {
  name: s__('SecurityReports|Scanner'),
  id: 'reportType',
  options: parseOptions(REPORT_TYPES),
  allOption: BASE_FILTERS.report_type,
  defaultOptions: [],
};

// This is used on the project-level report. It's used by the scanner filter that shows a list of
// scan types (DAST, SAST, etc) that's grouped by vendor.
export const vendorScannerFilter = {
  name: s__('SecurityReports|Scanner'),
  id: 'scanner',
  options: Object.keys(REPORT_TYPES).map((x) => createScannerOption(DEFAULT_SCANNER, x)),
  allOption: BASE_FILTERS.report_type,
  defaultOptions: [],
};

export const activityOptions = {
  NO_ACTIVITY: { id: 'NO_ACTIVITY', name: s__('SecurityReports|No activity') },
  WITH_ISSUES: { id: 'WITH_ISSUES', name: s__('SecurityReports|With issues') },
  NO_LONGER_DETECTED: { id: 'NO_LONGER_DETECTED', name: s__('SecurityReports|No longer detected') },
};

export const activityFilter = {
  name: s__('Reports|Activity'),
  id: 'activity',
  options: Object.values(activityOptions),
  allOption: BASE_FILTERS.activity,
  defaultOptions: [],
};

export const getProjectFilter = (projects) => {
  return {
    name: s__('SecurityReports|Project'),
    id: 'projectId',
    options: mapProjects(projects),
    allOption: BASE_FILTERS.project_id,
    defaultOptions: [],
  };
};

/**
 * Provided a security reports summary from the GraphQL API, this returns an array of arrays
 * representing a properly formatted report ready to be displayed in the UI. Each sub-array consists
 * of the user-friend report's name, and the summary's payload. Note that summary entries are
 * considered empty and are filtered out of the return if the payload is `null` or don't include
 * a vulnerabilitiesCount property. Report types whose name can't be matched to a user-friendly
 * name are filtered out as well.
 *
 * Take the following summary for example:
 * {
 *   containerScanning: { vulnerabilitiesCount: 123 },
 *   invalidReportType: { vulnerabilitiesCount: 123 },
 *   dast: null,
 * }
 *
 * The formatted summary would look like this:
 * [
 *   ['containerScanning', { vulnerabilitiesCount: 123 }]
 * ]
 *
 * Note that `invalidReportType` was filtered out as it can't be matched with a user-friendly name,
 * and the DAST report was omitted because it's empty (`null`).
 *
 * @param {Object} rawSummary
 * @returns {Array}
 */
export const getFormattedSummary = (rawSummary = {}) => {
  if (!isPlainObject(rawSummary)) {
    return [];
  }
  // Convert keys to snake case so they can be matched against REPORT_TYPES keys for translation
  const snakeCasedSummary = convertObjectPropsToSnakeCase(rawSummary);
  // Convert object to an array of entries to make it easier to loop through
  const summaryEntries = Object.entries(snakeCasedSummary);
  // Filter out empty entries as we don't want to display those in the summary
  const withoutEmptyEntries = summaryEntries.filter(
    ([, scanSummary]) => scanSummary?.vulnerabilitiesCount !== undefined,
  );
  // Replace keys with translations found in REPORT_TYPES if available
  const formattedEntries = withoutEmptyEntries.map(([scanType, scanSummary]) => {
    const name = REPORT_TYPES[scanType];
    return name ? [name, scanSummary] : null;
  });
  // Filter out keys that could not be matched with any translation and are thus considered invalid
  return formattedEntries.filter((entry) => entry !== null);
};

/**
 * We have disabled loading hasNextPage from GraphQL as it causes timeouts in database,
 * instead we have to calculate that value based on the existence of endCursor. When endCursor
 * is empty or has null value, that means that there is no next page to be loaded from GraphQL API.
 *
 * @param {Object} pageInfo
 * @returns {Object}
 */
export const preparePageInfo = (pageInfo) => {
  return { ...pageInfo, hasNextPage: Boolean(pageInfo?.endCursor) };
};

export const PROJECT_LOADING_ERROR_MESSAGE = __('An error occurred while retrieving projects.');

export default () => ({});
