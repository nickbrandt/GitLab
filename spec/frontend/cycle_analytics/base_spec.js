import { GlLoadingIcon, GlEmptyState } from '@gitlab/ui';
import { shallowMount } from '@vue/test-utils';
import Vue from 'vue';
import Vuex from 'vuex';
import { extendedWrapper } from 'helpers/vue_test_utils_helper';
import BaseComponent from '~/cycle_analytics/components/base.vue';
import PathNavigation from '~/cycle_analytics/components/path_navigation.vue';
import ValueStreamFilters from '~/cycle_analytics/components/value_stream_filters.vue';
import initState from '~/cycle_analytics/store/state';
import {
  permissions,
  transformedProjectStagePathData,
  selectedStage,
  convertedEvents as selectedStageEvents,
  createdBefore,
  createdAfter,
  currentGroup,
} from './mock_data';

const noDataSvgPath = 'path/to/no/data';
const noAccessSvgPath = 'path/to/no/access';

Vue.use(Vuex);

let wrapper;

const defaultState = {
  permissions,
  selectedStageEvents,
  selectedStage,
  currentGroup,
  createdBefore,
  createdAfter,
};

function createStore({ initialState = {}, initialGetters = {} }) {
  return new Vuex.Store({
    state: {
      ...initState(),
      ...defaultState,
      ...initialState,
    },
    getters: {
      pathNavigationData: () => transformedProjectStagePathData,
      ...initialGetters,
    },
  });
}

function createComponent({ initialState, initialGetters } = {}) {
  return extendedWrapper(
    shallowMount(BaseComponent, {
      store: createStore({ initialState, initialGetters }),
      propsData: {
        noDataSvgPath,
        noAccessSvgPath,
      },
    }),
  );
}

const findLoadingIcon = () => wrapper.findComponent(GlLoadingIcon);
const findPathNavigation = () => wrapper.findComponent(PathNavigation);
const findOverviewMetrics = () => wrapper.findByTestId('vsa-stage-overview-metrics');
const findStageTable = () => wrapper.findByTestId('vsa-stage-table');
const findEmptyStage = () => wrapper.findComponent(GlEmptyState);
const findStageEvents = () => wrapper.findByTestId('stage-table-events');
const findFilters = () => wrapper.findComponent(ValueStreamFilters);

describe('Value stream analytics component', () => {
  beforeEach(() => {
    wrapper = createComponent();
  });

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  it('renders the path navigation component', () => {
    expect(findPathNavigation().exists()).toBe(true);
  });

  it('renders the overview metrics', () => {
    expect(findOverviewMetrics().exists()).toBe(true);
  });

  it('renders the stage table', () => {
    expect(findStageTable().exists()).toBe(true);
  });

  it('renders the stage table events', () => {
    expect(findEmptyStage().exists()).toBe(false);
    expect(findStageEvents().exists()).toBe(true);
  });

  it('does not render the loading icon', () => {
    expect(findLoadingIcon().exists()).toBe(false);
  });

  describe('isLoading = true', () => {
    beforeEach(() => {
      wrapper = createComponent({
        initialState: { isLoading: true },
      });
    });

    it('renders the path navigation component with prop `loading` set to true', () => {
      expect(findPathNavigation().props('loading')).toBe(true);
    });

    it('does not render the overview metrics', () => {
      expect(findOverviewMetrics().exists()).toBe(false);
    });

    it('does not render the stage table', () => {
      expect(findStageTable().exists()).toBe(false);
    });

    it('renders the loading icon', () => {
      expect(findLoadingIcon().exists()).toBe(true);
    });
  });

  describe('isLoadingStage = true', () => {
    beforeEach(() => {
      wrapper = createComponent({
        initialState: { isLoadingStage: true },
      });
    });

    it('renders the stage table with a loading icon', () => {
      const tableWrapper = findStageTable();
      expect(tableWrapper.exists()).toBe(true);
      expect(tableWrapper.find(GlLoadingIcon).exists()).toBe(true);
    });

    it('renders the path navigation loading state', () => {
      expect(findPathNavigation().exists()).toBe(true);
    });
  });

  describe('isEmptyStage = true', () => {
    const emptyStageParams = {
      isEmptyStage: true,
      selectedStage: { ...selectedStage, emptyStageText: 'This stage is empty' },
    };
    beforeEach(() => {
      wrapper = createComponent({ initialState: emptyStageParams });
    });

    it('renders the empty stage with `Not enough data` message', () => {
      expect(findEmptyStage().html()).toMatchSnapshot();
    });

    describe('with a selectedStageError', () => {
      beforeEach(() => {
        wrapper = createComponent({
          initialState: {
            ...emptyStageParams,
            selectedStageError: 'There is too much data to calculate',
          },
        });
      });

      it('renders the empty stage with `There is too much data to calculate` message', () => {
        expect(findEmptyStage().html()).toMatchSnapshot();
      });
    });
  });

  describe('without enough permissions', () => {
    beforeEach(() => {
      wrapper = createComponent({
        initialState: {
          permissions: {
            ...permissions,
            [selectedStage.id]: false,
          },
        },
      });
    });

    it('renders the empty stage with `You need permission` message', () => {
      expect(findEmptyStage().html()).toMatchSnapshot();
    });
  });

  describe('without a selected stage', () => {
    beforeEach(() => {
      wrapper = createComponent({
        initialGetters: { pathNavigationData: () => [] },
        initialState: { selectedStage: null, isEmptyStage: true },
      });
    });

    it('renders the stage table', () => {
      expect(findStageTable().exists()).toBe(true);
    });

    it('renders the filters', () => {
      expect(findFilters().exists()).toBe(true);
    });

    it('does not render the path navigation', () => {
      expect(findPathNavigation().exists()).toBe(false);
    });

    it('does not render the stage table events', () => {
      expect(findStageEvents().exists()).toBe(false);
    });

    it('does not render the loading icon', () => {
      expect(findLoadingIcon().exists()).toBe(false);
    });
  });
});
