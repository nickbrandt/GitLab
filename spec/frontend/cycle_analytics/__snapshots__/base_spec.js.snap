// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Value stream analytics component isEmptyStage = true renders the empty stage with \`Not enough data\` message 1`] = `"<gl-empty-state-stub title=\\"We don't have enough data to show this stage.\\" svgpath=\\"path/to/no/data\\" description=\\"This stage is empty\\" class=\\"js-empty-state\\"></gl-empty-state-stub>"`;

exports[`Value stream analytics component isEmptyStage = true with a selectedStageError renders the empty stage with \`There is too much data to calculate\` message 1`] = `"<gl-empty-state-stub title=\\"There is too much data to calculate\\" svgpath=\\"path/to/no/data\\" description=\\"\\" class=\\"js-empty-state\\"></gl-empty-state-stub>"`;

exports[`Value stream analytics component without enough permissions renders the empty stage with \`You need permission\` message 1`] = `"<gl-empty-state-stub title=\\"You need permission.\\" svgpath=\\"path/to/no/access\\" description=\\"Want to see the data? Please ask an administrator for access.\\" class=\\"js-empty-state\\"></gl-empty-state-stub>"`;
